### PR TITLE
BUF_MEM grow should do memset only from old max in case of expansion

### PR DESCRIPTION
**Pull request by raja-ashok**
In SSL handshake, messages are kept in `BUF_MEM` structure variable `s->init_buf`. This is initially allocated for 16k and later shrunk or expanded. While receiving peer's message in `ssl3_get_message`, its shrunk(only length is reduced, not shrunk buf) to the received msg size. While sending message its expanded in some cases.

For example consider `ssl3_send_certificate_request`, here we are copying few data and then grow is called. If we written more data than `buf->length` before calling grow, then grow will reset it. **As it does memset from old length to new length.** 

So current BUF_MEM grow may cause some data lose. So below things are done in this PR.
1) After malloc memset `ret` for `n` bytes.
2) After realloc memset `ret` from max to `n`.
3) After shrink (`(str->length >= len)` case) memset from `len` to `str->length`
4) After expand (`(str->max >= len)` case) do not memset

### CODE REVIEW

**Comment by kroeckx** [comment]
This seems to turn BUF_MEM_grow() into BUF_MEM_grow_clean(). I don't think they're supposed to do the same. Maybe BUF_MEM_grow() shouldn't have any memset() calls in it?

File: crypto/buffer/buffer.c [hunk 1]
```diff
@@ -141,7 +145,6 @@ int BUF_MEM_grow_clean(BUF_MEM *str, size_t len)
         return (len);
     }
     if (str->max >= len) {
-        memset(&str->data[str->length], 0, len - str->length);
```
Comment (kroeckx) [inline crypto/buffer/buffer.c]: I'm not sure this is a good idea. Assuming you've always used the BUF_MEM_grow_clean() function (and no other bugs writing too much), that data should be all 0s. But BUF_MEM_new() doesn't do this for instance. I don't think you can rely on it already being 0 while the function promises to turn it into 0.

The same goes for the other change in this function, you try to optimize it by starting at the previous max instead of the previous length.

In short I think this whole function should be left as it was.

**Comment by kroeckx** [comment]
You also say that it's possible to write more than the length. I consider all cases where that happens to be a bug. Maybe it should be converted to use the WPACKET interface?

File: crypto/buffer/buffer.c (hunk 1 above)
Comment (raja-ashok) [inline crypto/buffer/buffer.c]: Hi kroeckx, Thanks for your comment. 

As per my knowledge, the different between `BUF_MEM_grow` and `BUF_MEM_grow_clean` is not providing memsetted buffer. Rather `BUF_MEM_grow_clean' should do `CRYPTO_realloc_clean`, that means during expansion it should do `OPENSSL_cleanse` before releasing old buffer.

So as per my knowledge both the function should return memsetted buffers.

**Comment by mattcaswell** [comment]
> If we written more data than buf->length before calling grow, then grow will reset it. As it does memset from old length to new length.

Maybe I'm missing the point - but I'm struggling to understand why this is a problem. So what if we do a memset from the old length the new length? If there is more data than buf->length then there shouldn't be. The condition that caused that to occur should be considered a bug - and that should be corrected. Removing the memset() will just mask that other bug - so I would consider this change detrimental.

**Comment by raja-ashok** [comment]
@mattcaswell : There are 2 ways of fixing.
1) Check and fix in the place where write is done more than buf->length. Like as you suggested.
2) In BUF_MEM grow memset buf from old max, after realloc

Out of these 2 options, I felt 2nd one is best. Because our only intension in BUF_MEM grow is to give memsetted buffer, I feel that we can do from old max, instead of old length after realloc. At same time need to memset complete buffer after malloc.

**Comment by mattcaswell** [comment]
This won't be merged due to considerations highlighted above.
